src/meta/meta_ckpt.c

```diff
@@ -1265,6 +1266,141 @@ __wt_meta_ckptlist_update_config(
     return (ret);
 }
 
+/*
+ * ckpt_blkmod_to_item --
+ *     Extracts the blkmod info in a checkpoint into a bitmap in a WT_ITEM.
+ */
+static int
+ckpt_blkmod_to_item(WT_SESSION_IMPL *session, WT_CKPT *ckpt, const char *id, WT_ITEM *output_item)
+{
+    WT_BLOCK_MODS *blk;
+    u_int i;
+    bool valid;
+
+    valid = false;
+    WT_CLEAR(*output_item);
+
+    for (i = 0, blk = &ckpt->backup_blocks[0]; i < WT_BLKINCR_MAX; ++i, ++blk)
+        if (F_ISSET(blk, WT_BLOCK_MODS_VALID))
+            valid = true;
+
+    /*
+     * If the existing block modifications are not valid, there is nothing to do.
+     */
+    if (!valid) {
+        return (0);
+    }
+
+    /*
+     * We have at least one valid modified block list.
+     */
+    for (i = 0, blk = &ckpt->backup_blocks[0]; i < WT_BLKINCR_MAX; ++i, ++blk) {
+        if (!F_ISSET(blk, WT_BLOCK_MODS_VALID))
+            continue;
+
+        if (strcmp(id, blk->id_str) == 0)
+            WT_RET(__wt_buf_set(session, output_item, blk->bitstring.data, blk->bitstring.size));
+    }
+
+    return (0);
+}
+
+/*
+ * get_blkmods --
+ *     Extracts the blkmod info from the metadata about a file+id into a bitmap in a WT_ITEM.
+ */
+static int
+get_blkmods(WT_SESSION_IMPL *session, const char *uri, const char *id, WT_ITEM *output_item)
+{
+    WT_CONFIG blkconf;
+    WT_CONFIG_ITEM blocks, key, value, blocks_key, blocks_value;
+    WT_CURSOR *metadata_cursor;
+    WT_DECL_RET;
+    char *file_config;
+
+    WT_CLEAR(blocks);
+    WT_CLEAR(key);
+    WT_CLEAR(value);
+    WT_CLEAR(blocks_key);
+    WT_CLEAR(blocks_value);
+    WT_CLEAR(*output_item);
+
+    metadata_cursor = NULL;
+    file_config = NULL;
+
+    WT_RET(__wt_metadata_cursor_open(session, NULL, &metadata_cursor));
```

Comment:
This function opens the metadata and positions at the uri.  But isn't it called from somewhere up the stack where we are positioned at the same entry in the metadata?  (I could be wrong). If that's the case, maybe the needed metadata can be passed down into this function rather than doing another lookup.

Reply:
I've had a look at the call stack and I don't think the metadata is always available to the calling code, but I may have missed something.

Reply:
Okay, thanks for checking.